File: frontend/src/types/tallyData.ts
export type TallySyncStepResult = {
  source_table: string;
  target_table: string;
  source_count: number;
  target_count_before: number;
  target_count_after: number;
  added: number;
  updated: number;
  unchanged: number;
  removed: number;
};

export type TallySyncSessionResponse = {
  started_at: string;
  completed_at: string;
  steps: TallySyncStepResult[];
  message: string;
};

File: backend/app/schemas/tally_data.py
"""Tally data sync request and response schemas."""

from __future__ import annotations

from datetime import UTC, datetime

from pydantic import BaseModel, Field, field_serializer


class TallySyncStepResult(BaseModel):
    """Result for one sync step in a session."""

    source_table: str
    target_table: str
    source_count: int
    target_count_before: int
    target_count_after: int
    added: int
    updated: int
    unchanged: int
    removed: int


class TallySyncSessionResponse(BaseModel):
    """Summary of a tally data sync session."""

    started_at: datetime
    completed_at: datetime
    steps: list[TallySyncStepResult] = Field(default_factory=list)
    message: str

    @field_serializer("started_at", "completed_at")
    def serialize_datetime(self, value: datetime) -> str:
        if value.tzinfo is None:
            value = value.replace(tzinfo=UTC)
        return value.isoformat()
